module/apmfasthttp/transaction.go

```diff
@@ -0,0 +1,84 @@
+package apmfasthttp
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/savsgio/gotils/bytes"
+	"go.elastic.co/apm"
+)
+
+var (
+	TxKey = fmt.Sprintf("__apmTxKey::%s__", bytes.Rand(make([]byte, 5)))
+
+	transactionPool = sync.Pool{
+		New: func() interface{} {
+			return new(Transaction)
+		},
+	}
+)
+
+func acquireTransaction() *Transaction {
+	return transactionPool.Get().(*Transaction)
+}
+
+func releaseTransaction(tx *Transaction) {
+	tx.reset()
+	transactionPool.Put(tx)
+}
+
+func (tx *Transaction) reset() {
+	tx.tracer = nil
+	tx.tx = nil
+	resetHTTPRequest(&tx.req)
+	tx.httpCtx = nil
+	tx.body = nil
+	tx.httpBody.reset()
+	tx.manualEnd = false
+}
+
+// AutoEnd defers the tx.End() to call it manually.
+//
+// So useful, if you need to response in streaming and count too that overtime.
+//
+// It is enabled by default.
+func (tx *Transaction) AutoEnd(v bool) {
+	tx.manualEnd = !v
+}
+
+// Tx returns the underlying APM transaction.
+func (tx *Transaction) Tx() *apm.Transaction {
+	return tx.tx
+}
+
+// Tx returns the underlying APM Tracer.
+func (tx *Transaction) Tracer() *apm.Tracer {
+	return tx.tracer
+}
+
+// End enqueues tx for sending to the Elastic APM server.
+//
+// Calling End will set tx's TransactionData field to nil, so callers
+// must ensure tx is not updated after End returns.
+//
+// If the underliying tx.Duration has not been set, End will set it to the elapsed time
+// since the transaction's start time.
+func (tx *Transaction) End() {
+	tx.tx.End()
+	releaseTransaction(tx)
```

Comment:
But the spans must be created before the transaction ends, right? If true, the transaction pool is safe because after the transaction ends you must not something more with the transaction and reduce the gc overhead.

Reply:
Your example case is similar to streaming response in `fasthttp`, and works well.

Reply:
```go
req := config.AcquireRequest()

if err := req.UnmarshalJSON(ctx.Request.Body()); err != nil {
  return errors.Wrap(err, "failed to unmarshal request config")
}

tx := apmfasthttp.GetTransaction(ctx) // Get the transaction from `RequestContext`
tx.AutoEnd(false) // set auto end to false, because you want to count the streaming write response

ctx.Response.SetBodyStreamWriter(func(w *bufio.Writer) {
   defer tx.End() // end the transaction

   // write in streaming...
})
```

Reply:
> But the spans must be created before the transaction ends, right? If true, the transaction pool is safe because after the transaction ends you must not something more with the transaction and reduce the gc overhead.

Spans may be started even after a transaction ends.